src/dune/super_context.ml

```diff
@@ -602,48 +602,73 @@ module Deps = struct
 
   let make_alias expander s =
     let loc = String_with_vars.loc s in
-    Expander.expand_path expander s |> Alias.of_user_written_path ~loc
+    Expander.Partial.expand_path expander s
+    |> String_with_vars.Partial.map ~f:(Alias.of_user_written_path ~loc)
+
+  let all_expanded xs ~f =
+    match
+      Result.List.fold_left xs ~init:[] ~f:(fun acc a ->
+          match f a with
+          | String_with_vars.Partial.Expanded a -> Ok (a :: acc)
+          | Unexpanded sw -> Error sw)
+    with
+    | Ok x -> String_with_vars.Partial.Expanded (List.rev x)
+    | Error sw -> Unexpanded sw
 
   let dep t expander = function
     | File s ->
-      let path = Expander.expand_path expander s in
-      let+ () = Build.path path in
-      [ path ]
+      Expander.Partial.expand_path expander s
+      |> String_with_vars.Partial.map ~f:(fun path ->
+             let+ () = Build.path path in
+             [ path ])
```

Comment:
I don't understand why we need to go through partial expansion here. The aim of partial expansion for actions is to be able to collect targets before the action is fully expanded. However, we never need this for the `deps` field given that it doesn't contain targets.

Reply:
We need to interpret `deps` before expanding targets. For example, consider how we evaluate something like this:

```
(action
 (deps (:foo %{bin:bar}))
 (target %{foo}))
```

Reply:
Agreed, but I still don't follow why we need to go through partial.

Reply:
The errors are discovered as we're doing the expansion. When such an error is discovered, the expansion of that variable is aborted and hence the only way to recover is through a partial expansion.

I suppose it's possible to make this error some other exception. It could be caught and handled instead of handling partial expansions.

Reply:
Ok, I feel like I'll need to go over all this code again at some point. I'm finding it difficult to follow everything. When you have left the airport chaos, we can maybe go over this together via videoconf.

Though, if someone else can review this don't wait for me.

Reply:
In case someone else is curious, here's the pseudo-code of the current code works:

```
String_with_vars.expand ~f:(fun var ->
  if binary_vailable var then
    add_failure resolved_forms;
    None (* This raises because we are in an [expand] *)
  else
    Some (binary_path)
)
```

My proposal is to change the above to use `partial_expand` over `expand`. This will prevent the `None` case from raising and we'll be able to extract the error from the resolved form.

To avoid partial expansion, I suggested changing `add_failure resolved_forms` to raise an exception. The caller can catch it and propagate it via `Build.fail`.

Reply:
Oh ok, I get it now!

In other places where we want to delay error reporting, such as for library name resolution, we are using the `Or_exn.t` monad. What do you think of doing the same here? That would make the intent explicit.

I was finding the code confusing because in my head partial is related to static vs dynamic. While here we are talking about delaying error reporting.

Reply:
Sure, I can do that. It would also make things a bit faster as we'd stop trying to expand the other variables.